test/k8sT/CustomCalls.go

```diff
@@ -0,0 +1,306 @@
+			localObjFile = filepath.Join(kubectl.BasePath(), "../bpf/custom", objFileName)
+			remoteObjFile = filepath.Join("/run/cilium/state", objFileName)
+
+			progPinPath = filepath.Join(bpffsDir, "cilium_bytecounter")
+			mapsPinDir = filepath.Join(bpffsDir, "cilium_bytecounter_maps")
+			mapPinPath = filepath.Join(mapsPinDir, "bytecount_map")
+
+			yaml = helpers.ManifestGet(kubectl.BasePath(), "demo-customcalls.yaml")
+			kubectl.ApplyDefault(yaml).ExpectSuccess("Unable to apply %s", yaml)
+
+			By("Compiling custom byte counter program")
+
+			err := kubectl.WaitForSinglePod(helpers.DefaultNamespace, compilerPodName, helpers.HelperTimeout)
+			Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("%s pod not ready after timeout", compilerPodName))
+
+			const bpfCustomDir string = "bpf/custom"
+
+			cmd := fmt.Sprintf("make -C %s clean V=0", bpfCustomDir)
+			res := kubectl.ExecPodCmd(helpers.DefaultNamespace, compilerPodName, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to clean up %s directory", bpfCustomDir))
+
+			cmd = fmt.Sprintf("make -C %s V=0", bpfCustomDir)
+			res = kubectl.ExecPodCmd(helpers.DefaultNamespace, compilerPodName, cmd)
+			res.ExpectSuccess("Failed to build custom byte counter program")
+		})
+
+		AfterAll(func() {
+			_ = kubectl.Delete(yaml)
+			ExpectAllPodsTerminated(kubectl)
+		})
+
+		checkOneDirection := func(endpointId int64, ciliumPod string,
+			clientName string, serverIP string, serverIdentity string,
+			direction customCallDirection) {
+
+			switch direction {
+			case IngressIPv4:
+				break
+			case EgressIPv4:
+				break
+			default:
+				// IPv6 packets do not have the same length,
+				// and the comparison on the counter value will
+				// fail. Could be adjusted in the future.
+				Failf("Direction not supported: %v", direction)
+			}
+
+			By("Updating call map with reference to custom program")
+
+			// TODO: other directions
+			cmd := fmt.Sprintf("bpftool map update pinned %scilium_calls_custom_%05d key %d 0 0 0 value pinned %s", bpffsDir, endpointId, direction, progPinPath)
+			res := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to update call map with reference to custom program ('%s'): %s", cmd, res.Stderr()))
+
+			By("Sending traffic between the pods")
+
+			cmd = helpers.Ping(serverIP)
+			res = kubectl.ExecPodCmd(helpers.DefaultNamespace, clientName, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to ping from %s to %s", "", ""))
+
+			By("Retrieving counter value")
+
+			cmd = fmt.Sprintf("bpftool -j map lookup pinned %s key %s | jq -r '.value[]'", mapPinPath, serverIdentity)
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to lookup byte counter value ('%s'): %s", cmd, res.Stderr()))
+
+			// Output from bpftool is on the form
+			// "0x11\n0x22\n0x33\n...0x88\n", convert it.
+			// Too bad bpftool returns hex that jq doesn't know how
+			// to convert yet.
```

Comment:
(Completely unrelated to this PR) Definitely feels like there's a gap in the overall JSON workflow here if you can't just pipe bpftool into jq to do the operations you want on the data. Maybe BTF can solve this for us some day.

Reply:
My fault, actually :( I wasn't much experienced with JSON when I added it to bpftool, and I had the terrible idea of making it print hexa strings. There is a built-in function in jq (`tonumber`) to parse strings as integers but it doesn't support hex strings yet (I found an open issue for that).

I'm thinking of adding an option (a new one, or maybe double `-jj`|`-pp`) to have bpftool dump real integers. Feels dirty, but I guess hex strings are UAPI by kernel's standards now :(.

Reply:
Given that the values are arbitrary length and often make the most sense as hexadecimal if they can't otherwise be better-interpreted, I can understand avoiding encoding directly as JSON integer types since the common display format would likely render them in decimal. Is bpftool output format really subject to kernel UAPI stability though?

Reply:
That was my understanding. Maybe the maintainers would be more flexible than for traditional kernel UAPI, but it's hard to say. Changing it would likely break some scripts. That would be a discussion for bpf@vger I suppose.